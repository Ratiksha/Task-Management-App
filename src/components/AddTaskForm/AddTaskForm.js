import React, { Component } from 'react';
import './AddTaskForm.scss';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";

export default class AddTaskForm extends Component{
    constructor(props){
        super(props);
        this.state = {
            value: '',
            data: '',
            isAdding: false,
            errorRequired: false,
            errorDuplicate: false,
            disableSubmit: true
        }
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange = (event) => {
        this.setState({
            value: event.target.value, 
            errorRequired:false, 
            errorDuplicate: false,
            disableSubmit: false
        });
    }

    handleSubmit = (event) => {
        const { addTask, listId, editTask, hideModal, taskId } = this.props
        const { value } = this.state
        event.preventDefault();
        if(!this.checkDuplicate(value)){
        if(value && addTask){
            addTask(value, listId);
            this.setEditing(false);
        }
        else if(value && editTask){
            editTask(value, listId, taskId);
            hideModal();
        }
        else{
            this.setState({
                errorRequired: true
            })
        }
    }
    }

    checkDuplicate = (name) => {
        const lists = this.props.lists ? this.props.lists : ''
        for (var i = 0; i < lists.length; i++) {
                for (var j = 0; j < lists[i].cards.length; j++) {
                    if(lists[i].cards[j].taskName.toUpperCase() === name.toUpperCase()){
                        this.setState({
                            errorDuplicate: true
                        })
                        return true
                    }
                }
          }
          
    }

    setEditing = (isAdding) => {
        this.setState({
            isAdding,
            errorRequired:false, 
            errorDuplicate: false,
            value: ''
        });
      }
    
    render() {
        const { isAdding, value, errorRequired, errorDuplicate, disableSubmit } = this.state;
        const { addTask, editTask, name } = this.props;
        if(!isAdding && addTask){
            return(
                <div className="add-button">
                     <button onClick={() => this.setEditing(true)}> <FontAwesomeIcon icon={faPlus}/> Add Task</button>
                </div>
            )
           
        }
        return (
            <form onSubmit={this.handleSubmit}>
                {editTask && <h3>Edit Task</h3>}
                <input type="text" defaultValue={name? name : value} onChange={this.handleChange} className={`${errorRequired || errorDuplicate? 'highlight': ''}`} />
                <input type="submit" value="Submit" className="button green" disabled={editTask ? disableSubmit : false} />
                {errorRequired && <p className="error">Please enter the name</p>}
                {errorDuplicate && <p className="error">Task already exists</p>}
                {!editTask && <button onClick={() => this.setEditing(false)} className="grey">Cancle</button>}
            </form>
        )
    }
}