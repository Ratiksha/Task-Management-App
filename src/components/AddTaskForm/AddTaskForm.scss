.add-button button{
    background: var(--color);
    margin-top: 10px;
}