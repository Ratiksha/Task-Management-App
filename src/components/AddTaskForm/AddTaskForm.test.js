import React from 'react';
import { shallow } from 'enzyme';
import AddTaskForm from './AddTaskForm';

let wrapped = shallow(<AddTaskForm />);
describe('AddTaskForm', () => {
  it('should render the AddTaskForm Component correctly', () => {   
    expect(wrapped).toMatchSnapshot();
  });

  it('When submit button is clicked', () => {
    const preventDefault = jest.fn();
    wrapped.find('form').simulate('submit', { preventDefault })
    expect(preventDefault).toBeCalled();
  });

  it('When input is clicked', () => {
    const mockedEvent = { target: { value: "foo"} }
    const spy = jest.spyOn(wrapped.instance(), 'handleChange')
    wrapped.instance().forceUpdate();
    wrapped.find('input[type="text"]').simulate('change', mockedEvent)
    expect(spy).toHaveBeenCalledTimes(1);
  })
});