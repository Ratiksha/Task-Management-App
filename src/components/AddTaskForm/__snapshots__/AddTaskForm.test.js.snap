// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm should render the AddTaskForm Component correctly 1`] = `ShallowWrapper {}`;
